dev-packages/e2e-tests/lib/getTestMatrix.ts

```diff
@@ -0,0 +1,151 @@
+import { execSync } from 'child_process';
+import * as fs from 'fs';
+import * as path from 'path';
+import { sync as globSync } from 'glob';
+
+interface MatrixInclude {
+  'test-application': string;
+  'build-command'?: string;
+  'assert-command'?: string;
+  label?: string;
+}
+
+/**
+ * This methods generates a matrix for the GitHub Actions workflow to run the E2E tests.
+ * It checks which test applications are affected by the current changes in the PR and then generates a matrix
+ * including all test apps that have at least one dependency that was changed in the PR.
+ * If no `--base=xxx` is provided, it will output all test applications.
+ *
+ * If `--optional=true` is set, it will generate a matrix of optional test applications only.
+ * Otherwise, these will be skipped.
+ */
+function run(): void {
+  const args: Record<string, string> = {};
+  process.argv
+    .slice(2)
+    .filter(arg => arg.startsWith('--') && arg.includes('='))
+    .forEach(arg => {
+      const [part1, part2] = arg.split('=') as [string, string];
+      const argName = part1.replace('--', '');
+      const argValue = part2;
+      args[argName] = argValue;
+    });
+
+  const { base, head, optional = 'false' } = args;
+
+  const testApplications = globSync('*', { cwd: `${__dirname}/../test-applications/` });
```

Comment:
We could be a bit more clever and glob for `test-applications/*/package.json` right away so we don't have to do the dance with trying to find the package json.